sources/transaction.move

```diff
@@ -1,2 +1,130 @@
 module bitcoin_spv::transaction;
+use bitcoin_spv::btc_math::{btc_hash, covert_to_compact_size, to_number, compact_size};
+use bitcoin_spv::utils::slice;
 
+// === BTC script opcode ===
+const OP_DUP: u8= 0x76;
+const OP_HASH160: u8 = 0xa9;
+const OP_DATA_20: u8 = 0x14;
+const OP_EQUALVERIFY: u8  = 0x88;
+const OP_CHECKSIG: u8 = 0xac;
+
+// public struct Input has copy, drop {
+//     tx_id: vector<u8>,
+//     vout: vector<u8>,
+//     script_size: u256,
+//     script_sig: vector<u8>,
+//     sequece: vector<u8>,
+// }
+
+public struct Output has copy, drop {
+    amount: u256,
+    script_pubkey: vector<u8>
+}
+
+public struct Transaction has copy, drop {
+    version: vector<u8>,
+    input_count: u256,
+    inputs: vector<u8>,
+    output_count: u256,
+    outputs: vector<Output>,
+    tx_id: vector<u8>,
+    lock_time: vector<u8>
+}
+
+
+// TODO: better name for this.
+// we don't create any new transaction
+public fun new_transaction(
+    version: vector<u8>,
+    input_count: u256,
+    inputs: vector<u8>,
+    output_count: u256,
+    outputs: vector<u8>,
+    lock_time: vector<u8>,
+): Transaction {
+    assert!(version.length() == 4);
+    assert!(lock_time.length() == 4);
+    let number_input_bytes = covert_to_compact_size(input_count);
+    let number_output_bytes = covert_to_compact_size(output_count);
+
+    // compute TxID
+    let mut tx_data = version;
+    tx_data.append(number_input_bytes);
+    tx_data.append(inputs);
+    tx_data.append(number_output_bytes);
+    tx_data.append(outputs);
+    tx_data.append(lock_time);
+
+    let tx_id = btc_hash(tx_data);
+
+    let outputs_decoded = decode_outputs(output_count, outputs);
+
+    Transaction {
+        version,
+        input_count,
+        inputs,
+        output_count,
+        outputs: outputs_decoded,
+        lock_time,
+        tx_id
+    }
+}
+
+public fun new_output(amount: u256, script_pubkey: vector<u8>): Output {
+    Output {
+        amount,
+        script_pubkey
+    }
+}
+
+public fun tx_id(tx: &Transaction): vector<u8> {
```

Comment:
maybe we should calcualte it on demand? only when its needed? rather than calculating it already in the `new_transaction`?

Reply:
This computation is heavy, and every time we create a transaction, we always want to use tx_id.
This way, I compute this one time more :D We can have a few optimal for this action. However we should do the simple version in the begin I think